docs/src/advanced_topics/execution_trace_optimization.md

```diff
@@ -0,0 +1,44 @@
+# Execution trace optimization
+
+## Understanding cycle counts in Miden VM
+
+When we refer to "number of cycles" in most Miden VM documentation, we're specifically referring to the **stack rows** portion of the execution trace. However, the actual proving time is determined by what we call the "true number of cycles," which is the maximum of all trace segment lengths:
+
+- **Stack rows**: One row per VM operation (what `clk` outputs)
+- **Range checker rows**: Added when opcodes request 16-bit range checks
+- **Chiplet rows**: Added when opcodes call specialized chiplets
+
+The **true number of cycles** = `max(stack_rows, range_checker_rows, chiplet_rows)`
+
+In some cases, either the range checker or chiplets could end up requiring more rows than the stack rows, making the true cycle count higher than what the stack-based cycle counter reports.
+
+## Analyzing trace segments with miden-vm analyze
+
+The `miden-vm analyze` command provides detailed information about trace segment utilization, showing:
+- Stack rows used
+- Range checker rows used  
+- Chiplet rows used
+- The resulting true number of cycles (maximum of the three)
+
+This tool helps identify which trace segments are driving the overall proving time for a given program.
+
+## Trace segment growth and proving performance
+
+Even when two programs run the same number of VM cycles, their proving time can differ significantly because of how the execution trace is structured.
+
+| Trace segment      | Purpose                                              | Native growth rule                                     |
+| ------------------ | ---------------------------------------------------- | ------------------------------------------------------ |
+| Stack rows         | Core transition constraints; one row per opcode      | +1 row for every operation                             |
+| Range‑checker rows | Ensure selected values lie in \[0 .. 2¹⁶)            | Rows added only when an opcode requests a 16‑bit check |
+| Chiplet rows       | Bitwise, hash, memory and other accelerator circuits | Rows added only when an opcode calls a chiplet         |
+
+1. **Independent growth**
+   Each segment expands on its own. A pure arithmetic loop inflates only the stack segment, whereas repeated hashing inflates the chiplet segment.
+
+2. **Power‑of‑two padding**
+   After execution halts, the prover finds the largest trace segment length `L`, rounds it up to the next power of two `Lʹ = 2^ceil(log₂ L)`, and pads all rows with zeros until all trace segments reach `Lʹ`. The AIR expects a square trace matrix of this size.
```

Comment:
```suggestion
   After execution halts, the prover finds the largest trace segment length `L`, rounds it up to the next power of two `Lʹ = 2^ceil(log₂ L)`, and pads all rows with zeros until all trace segments reach `Lʹ`. The prover expects a square trace matrix of this size.
```
Although `AIR` is correct, I think using "prover" is a bit more intuitive for people